amp/parse.go

```diff
@@ -32,6 +41,114 @@ type Size struct {
+)
+
+// ReadPolicy returns a privacy writer in accordance to the query values consent, consent_type and gdpr_applies.
+// Returned policy writer could either be GDPR, CCPA or NilPolicy. The second return value is a warning.
+func ReadPolicy(ampParams Params, pbsConfigGDPREnabled bool) (privacy.PolicyWriter, error) {
+	if len(ampParams.Consent) == 0 {
+		return privacy.NilPolicyWriter{}, nil
+	}
+
+	var rv privacy.PolicyWriter = privacy.NilPolicyWriter{}
+	var warning error
+	var warningMsg string
+
+	// If consent_type was set to CCPA or GDPR TCF2, we return a valid writer even if the consent string is invalid
+	switch ampParams.ConsentType {
+	case ConsentTCF1:
+		warningMsg = "TCF1 consent is deprecated and no longer supported."
+	case ConsentTCF2:
+		if pbsConfigGDPREnabled {
+			rv = buildGdprTCF2ConsentWriter(ampParams)
+			// Log warning if GDPR consent string is invalid
+			warningMsg = validateTCf2ConsentString(ampParams.Consent)
+		}
+	case ConsentUSPrivacy:
+		rv = ccpa.ConsentWriter{ampParams.Consent}
+		if ccpa.ValidateConsent(ampParams.Consent) {
+			if parseGdprApplies(ampParams.GdprApplies) == 1 {
+				// Log warning because AMP request comes with both a valid CCPA string and gdpr_applies set to true
+				warningMsg = "AMP request gdpr_applies value was ignored in account of provided consent string found to be CCPA and not GDPR."
+			}
+		} else {
+			// Log warning if CCPA string is invalid
+			warningMsg = fmt.Sprintf("Consent string '%s' is not a valid CCPA consent string.", ampParams.Consent)
+		}
+	case ConsentNone:
+		fallthrough
+	default:
+		if ccpa.ValidateConsent(ampParams.Consent) {
+			rv = ccpa.ConsentWriter{ampParams.Consent}
+			if parseGdprApplies(ampParams.GdprApplies) == 1 {
+				warningMsg = "AMP request gdpr_applies value was ignored in account of provided consent string found to be CCPA and not GDPR."
+			}
+		} else if pbsConfigGDPREnabled && len(validateTCf2ConsentString(ampParams.Consent)) == 0 {
+			rv = buildGdprTCF2ConsentWriter(ampParams)
+		} else {
+			warningMsg = fmt.Sprintf("Consent '%s' is not recognized as CCPA nor GDPR TCF2.", ampParams.Consent)
+		}
+	}
+
+	if len(warningMsg) > 0 {
+		warning = &errortypes.Warning{
+			Message:     warningMsg,
+			WarningCode: errortypes.InvalidPrivacyConsentWarningCode,
+		}
+	}
+	return rv, warning
+}
+
+// buildGdprTCF2ConsentWriter returns a gdpr.ConsentWriter that will set regs.ext.gdpr to the value
+// of 1 if gdpr_applies wasn't defined. The reason for this is that this function gets called when
+// GDPR applies, even if field gdpr_applies wasn't set in the AMP endpoint query.
+func buildGdprTCF2ConsentWriter(ampParams Params) gdpr.ConsentWriter {
+	writer := gdpr.ConsentWriter{Consent: ampParams.Consent}
+
+	// If gdpr_applies was not set, regs.ext.gdpr must equal 1
+	var gdprValue int8 = 1
+	if ampParams.GdprApplies != nil {
+		// set regs.ext.gdpr if non-nil gdpr_applies was set to true
+		gdprValue = parseGdprApplies(ampParams.GdprApplies)
+	}
```

Comment:
> here it appears we default to a gdprValue of 1 if it is nil and in the other cases we are defaulting to a value of 0 since the nil check in the function will return 0. Is that correct?

Correct. That's the reason I added the `nil` check here and not in the other instances. From the specs and previous discussions we agreed that if `gdpr_applies` was not set, the `gdpr.ConsentWriter` must default to write `regs.ext.gdpr` with a value of 1. On the other hand, if `gdpr_applies` was set/non-nil, the `gdpr.ConsentWriter` must write into `regs.ext.gdpr` whatever value `gdpr_applies` holds.